tensorrt_llm/serve/scheduler.py

```diff
@@ -0,0 +1,217 @@
+        super().__init__(ctx_servers, gen_servers)
+        # Load map between servers and their number of tokens processed
+        self._ctx_server_load = {}
+        self._gen_server_load = {}
+
+        self._ctx_server_load_heap = []
+        self._gen_server_load_heap = []
+
+        # Routing table to map requests to servers
+        self._ctx_req_routing_table = {}
+        self._gen_req_routing_table = {}
+
+        self._init_heap()
+
+    def _init_heap(self):
+        if self._ctx_servers is not None:
+            for server in self._ctx_servers:
+                self._ctx_server_load[server] = 0
+                heapq.heappush(self._ctx_server_load_heap,
+                               (self._ctx_server_load[server], server))
+
+        if self._gen_servers is not None:
+            for server in self._gen_servers:
+                self._gen_server_load[server] = 0
+                heapq.heappush(self._gen_server_load_heap,
+                               (self._gen_server_load[server], server))
+
+    def _get_request_num_tokens(self, request: CompletionRequest,
+                                server_type: str) -> int:
+        if server_type == "context":
+            if isinstance(request.prompt, str) or \
+                (isinstance(request.prompt, list) and isinstance(request.prompt[0], int)):
+                prompts = [request.prompt]
+            else:
+                prompts = request.prompt
+
+            num_tokens = sum(len(prompt) for prompt in prompts)
+        else:
+            num_tokens = 1
+        return num_tokens
+
+    def schedule(self, request: CompletionRequest, server_type: str) -> str:
+        num_tokens = self._get_request_num_tokens(request, server_type)
+
+        with self._lock:
+            if server_type == "context":
+                server = heapq.heappop(self._ctx_server_load_heap)[1]
+                self._ctx_server_load[server] += num_tokens
+                heapq.heappush(self._ctx_server_load_heap,
+                               (self._ctx_server_load[server], server))
+                # TODO: Can we have a unique id for each request?
+                self._ctx_req_routing_table[request.prompt] = server
+            else:
+                server = heapq.heappop(self._gen_server_load_heap)[1]
+                self._gen_server_load[server] += num_tokens
+                heapq.heappush(self._gen_server_load_heap,
+                               (self._gen_server_load[server], server))
+                # TODO: Can we have a unique id for each request?
+                self._gen_req_routing_table[request.prompt] = server
+        return server
+
+    def finish_request(self, request: CompletionRequest, server_type: str):
+        num_tokens = self._get_request_num_tokens(request, server_type)
+
+        with self._lock:
+            if server_type == "context":
+                server = self._ctx_req_routing_table[request.prompt]
+                self._ctx_server_load[server] -= num_tokens
```

Comment:
As suggested earlier, we should have a `decrement_load` function for the `ServerState`, and store a map between `server`s and `ServerState`s